### PR TITLE
Code to generate legend graphic with some of the export ui code

### PR DESCRIPTION
**Pull request by AleksueiR**
#### Limitations:
- generation crashes if some layers don't have names
- generation crashes if some layers failed to load
- generation crashes if some layers don't have symbology defined (didn't load yet)
- long layer names without spaces cannot be wrapped right now
- low section count might crash the generation
- small legends looks weird spread out over many sections
- it sort of works in IE
- lots of typos in the code

Relates #1176

---
This change is [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1254)

### CODE REVIEW

**Comment by alyec** [comment]
Reviewed 12 of 12 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/app/ui/ui.module.js, line 9 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTExwydCCIw27JI0mRn:-KTExwyePdMe6AwwILwt:b-un9ebt) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/ui.module.js#L9)):*
> ```JavaScript
>      *
>      * The `app.ui` module pull in all the inidividual ui modules.
>      */
> ```

Can we add a note about flattening the UI modules here?  If it's easy maybe we should flatten the new export module.

---

*[src/app/ui/export/export-legend.service.js, line 36 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF5zIL1u_IcfBZmA0n:-KTF5zIMg5OEmtWoqCun:b5rf4km) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L36)):*
> ```JavaScript
>     function exportLegendService($rootElement, geoService) {
>         const service = {
>             generate
> ```

All functions besides generate should be marked `@private`

---

*[src/app/ui/export/export-legend.service.js, line 44 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF6um_b5tgDmpnBOr3:-KTF6um_b5tgDmpnBOr4:bk02fi5) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L44)):*
> ```JavaScript
> 
>         /**
>          * @function generate
> ```

Missing description

---

*[src/app/ui/export/export-legend.service.js, line 60 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTEzEJxaUvP0xW_prWZ:-KTEzEJxaUvP0xW_prW_:bzh8s2c) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L60)):*
> ```JavaScript
>             $rootElement.append(hiddenNode);
> 
>             var legend = SVG(hiddenNode[0]).size(availableWidth, 100);
> ```

I think every var / let declaration in here can become a const

---

*[src/app/ui/export/export-legend.service.js, line 93 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF-i2Aeq2Gi2HQoXon:-KTF-i2Aeq2Gi2HQoXoo:b-p8786o) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L93)):*
> ```JavaScript
>          * @return {Number} calculated section height
>          */
>         function findOptimumSectionHeight(legi, sectionCount) {
> ```

Can you summarize the algorithm either in the header comment or within the function?

---

*[src/app/ui/export/export-legend.service.js, line 141 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTEzvn4l-h9Pnt2ZYKB:-KTEzvn4l-h9Pnt2ZYKC:b-pxwlkx) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L141)):*
> ```JavaScript
>                     }
> 
>                     if (iteration > 42) {
> ```

An excellent choice of arbitrary constants

---

*[src/app/ui/export/export-legend.service.js, line 191 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF12qBCueP7vs_fpJ2:-KTF12qBCueP7vs_fpJ3:b-umaufn) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L191)):*
> ```JavaScript
>                         const [lineX, lineY, lineHeight] = [line.x(), line.y(), line.height()];
> 
>                         // if the line starts below the wrapping elemenet, skip
> ```

typo: element

---

*[src/app/ui/export/export-legend.service.js, line 204 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF19urswDOmEaKse9M:-KTF19urswDOmEaKse9N:b-2uv24g) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L204)):*
> ```JavaScript
>                                 .stroke('black');
> 
>                             line.remove(); // remove oribinal line
> ```

typo: original

---

*[src/app/ui/export/export-legend.service.js, line 274 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF5RjkWYMNS7W0e5Sc:-KTF5RjlqmP8tz7_M9Ff:b34og8a) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L274)):*
> ```JavaScript
> 
>             /**
>              * Makes stuff.
> ```

The department of redundancy department would like to award you with it's excellent award of excellence.  :trophy: 

---

*[src/app/ui/export/export.service.js, line 1 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF7aToBzlEsCDRqXMW:-KTF7aToBzlEsCDRqXMX:bxfssj9) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export.service.js#L1)):*
> ```JavaScript
> (() => {
> ```

missing jsdoc headers on the functions

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by james-rae** [comment]
Reviewed 10 of 12 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by AleksueiR** [comment]
Review status: 8 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/app/ui/export/export-legend.service.js, line 36 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF5zIL1u_IcfBZmA0n:-KTFP74Xdl4svdFv9Ynt:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L36)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> All functions besides generate should be marked `@private`
</details>
Done.

---

*[src/app/ui/export/export-legend.service.js, line 44 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF6um_b5tgDmpnBOr3:-KTFPFzit9bOpK74y95E:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L44)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Missing description
</details>
Done.

---

*[src/app/ui/export/export-legend.service.js, line 60 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTEzEJxaUvP0xW_prWZ:-KTF8ah_q3kixEYYcNPN:bqadrys) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L60)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> I think every var / let declaration in here can become a const
</details>
`sectionCount` is let because it can change (not implemented yet). Changed others.

---

*[src/app/ui/export/export-legend.service.js, line 93 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF-i2Aeq2Gi2HQoXon:-KTFSFoeu1yswd3GDu6d:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L93)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Can you summarize the algorithm either in the header comment or within the function?
</details>
Done.

---

*[src/app/ui/export/export-legend.service.js, line 141 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTEzvn4l-h9Pnt2ZYKB:-KTF9U0IiDuySBapmoze:b1f53gd) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L141)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> An excellent choice of arbitrary constants
</details>


---

*[src/app/ui/export/export-legend.service.js, line 191 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF12qBCueP7vs_fpJ2:-KTFSK5N7nsDGU_gd-jD:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L191)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> typo: element
</details>
Done.

---

*[src/app/ui/export/export-legend.service.js, line 204 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF19urswDOmEaKse9M:-KTFSLh9XkJrUPjDAzWM:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L204)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> typo: original
</details>
Done.

---

*[src/app/ui/export/export-legend.service.js, line 274 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF5RjkWYMNS7W0e5Sc:-KTFSNJcitq_cllal9Zb:b-ge1163) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export-legend.service.js#L274)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> The department of redundancy department would like to award you with it's excellent award of excellence.  :trophy: 
</details>
Accepted!

---

*[src/app/ui/export/export.service.js, line 1 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTF7aToBzlEsCDRqXMW:-KTFSSA94iTaxq8ABbUy:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/export/export.service.js#L1)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> missing jsdoc headers on the functions
</details>
Done.

---

*[src/app/ui/ui.module.js, line 9 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTExwydCCIw27JI0mRn:-KTFOXSj7X8tbGBXn-9p:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/c659ee330aaf417eb18d307efefafab8c22d409b/src/app/ui/ui.module.js#L9)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Can we add a note about flattening the UI modules here?  If it's easy maybe we should flatten the new export module.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by alyec** [comment]
Reviewed 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/app/ui/export/export-legend.service.js, line 91 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTFeltA86bdeIoKOfS0:-KTFeltA86bdeIoKOfS1:b-bjq0c) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/dec227725dfe832a70f918f5ef7710e7a316e5e9/src/app/ui/export/export-legend.service.js#L91)):*
> ```JavaScript
>          * Initially, the legend is constructed as a single list (section) and this function determines the best places to wrap it. It iterates over legend items keeping a running height. When the running height is greater than the height of the section, the element straddling the boundary )(with items following it) is wrapped (moved to a new section). Since that element was removed from the section, the total height of remaining elements is less than the section height and the difference is added to the running height.
>          *
>          * If after splitting the legend into `sectionCount` number of columns, the running height of the legend is greater than `sectionCount x sectionHeight` it means the legend doesn't fit. The section height is increased by 1px and tested again. After the minimum section height is found, do forty two more iterations increasing the section height. After that, pick the section hight which results is smallest amount of whitespace. Whitespace is created when large legend elements (wms images) are wrapped.
> ```

typo: section height (third sentence)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by AleksueiR** [comment]
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/app/ui/export/export-legend.service.js, line 91 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254#-KTFeltA86bdeIoKOfS0:-KTFfq17-U5ff68sVRm-:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/dec227725dfe832a70f918f5ef7710e7a316e5e9/src/app/ui/export/export-legend.service.js#L91)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> typo: section height (third sentence)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by james-rae** [comment]
Reviewed 1 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*

**Comment by alyec** [comment]
Reviewed 1 of 1 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1254)*
